File: source/includes/aggregation.scala
import org.mongodb.scala._
import org.mongodb.scala.model.{ Aggregates, Filters, Accumulators }
import org.mongodb.scala.bson.Document
import com.mongodb.ExplainVerbosity
import org.mongodb.scala.model.Projections
import org.mongodb.scala.model.search._
import org.mongodb.scala.model.search.SearchOptions.searchOptions

object Aggregation {

  def main(args: Array[String]): Unit = {
    val mongoClient = MongoClient("<connection string URI>")

    val database: MongoDatabase = mongoClient.getDatabase("sample_restaurants")
    val collection: MongoCollection[Document] = database.getCollection("restaurants")

    // Retrieves documents with a cuisine value of "Bakery", groups them by "borough", and
    // counts each borough's matching documents
    // start-match-group
    val pipeline = Seq(Aggregates.filter(Filters.equal("cuisine", "Bakery")),
                       Aggregates.group("$borough", Accumulators.sum("count", 1))
    )

    collection.aggregate(pipeline)
              .subscribe((doc: Document) => println(doc.toJson()),
                        (e: Throwable) => println(s"There was an error: $e"))
    // end-match-group

    // Performs the same aggregation operation as above but asks MongoDB to explain it
    // start-explain
    val pipelineToExplain = Seq(Aggregates.filter(Filters.equal("cuisine", "Bakery")),
                       Aggregates.group("$borough", Accumulators.sum("count", 1))
    )

    collection.aggregate(pipelineToExplain)
              .explain(ExplainVerbosity.EXECUTION_STATS)
              .subscribe((doc: Document) => println(doc.toJson()),
                        (e: Throwable) => println(s"There was an error: $e"))
    // end-explain

    // start-atlas-search
    val operator = SearchOperator.text(SearchPath.fieldPath("name"), "Salt")
    val options = searchOptions().index("<search index name>")
    
    val pipeline = Seq(Aggregates.search(operator, options),
                       Aggregates.project(Projections.include("name")))

    collection.aggregate(pipeline)
              .subscribe((doc: Document) => println(doc.toJson()),
                        (e: Throwable) => println(s"There was an error: $e"))
    // end-atlas-search

    Thread.sleep(1000)
    mongoClient.close()
  }
}

File: content/cpp/upcoming/source/includes/aggregation.cpp
#include <iostream>

#include <bsoncxx/builder/basic/document.hpp>
#include <bsoncxx/builder/basic/kvp.hpp>
#include <bsoncxx/json.hpp>
#include <mongocxx/client.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/pipeline.hpp>
#include <mongocxx/uri.hpp>

using bsoncxx::builder::basic::kvp;
using bsoncxx::builder::basic::make_document;

int main() {
    mongocxx::instance instance;
    mongocxx::uri uri("<connection string>");
    mongocxx::client client(uri);

    auto db = client["sample_restaurants"];
    auto collection = db["restaurants"];

    {
        // Retrieves documents with a cuisine value of "Bakery", groups them by "borough", and
        // counts each borough's matching documents
        // start-match-group
        mongocxx::pipeline stages;

        stages.match(make_document(kvp("cuisine", "Bakery")))
              .group(make_document(kvp("_id", "$borough"), kvp("count", make_document(kvp("$sum", 1)))));

        auto cursor = collection.aggregate(stages);

        for (auto&& doc : cursor) {
            std::cout << bsoncxx::to_json(doc) << std::endl;
        }
        // end-match-group
    }

    {
        // Performs the same aggregation operation as above but asks MongoDB to explain it
        // start-explain
        mongocxx::pipeline stages;

        stages.match(make_document(kvp("cuisine", "Bakery")))
              .group(make_document(kvp("_id", "$borough"), kvp("count", make_document(kvp("$sum", 1)))));

        auto command = make_document(
            kvp("explain", make_document(
                                kvp("aggregate", "restaurants"),
                                kvp("pipeline", stages.view_array()),
                                kvp("cursor", make_document()))));

        auto result = db.run_command(command.view());
        std::cout << bsoncxx::to_json(result) << std::endl;
        // end-explain
    }
}